File: 073SameBSTs/same_bsts.rb
# Solution 1
# O(n^2) time / O(n^2) space

def same_bsts(array_one, array_two)
  if array_one.length != array_two.length
    return false
  end

  if array_one.length == 0 && array_two.length == 0
    return true
  end

  if array_one[0] != array_two[0]
    return false
  end

  left_one = get_smaller(array_one)
  left_two = get_smaller(array_two)
  right_one = get_bigger_or_equal(array_one)
  right_two = get_bigger_or_equal(array_two)

  return same_bsts(left_one, left_two) && same_bsts(right_one, right_two)
end

def get_smaller(arr)
  smaller = []
  (1..arr.length - 1).each do |i|
    if arr[i] < arr[0]
      smaller << arr[i]
    end
  end
  return smaller
end

def get_bigger_or_equal(arr)
  bigger_or_equal = []
  (1..arr.length - 1).each do |i|
    if arr[i] >= arr[0]
      bigger_or_equal << arr[i]
    end
  end
  return bigger_or_equal
end

# Solution 2
# O(n^2) time / O(d) space
# d - depth of BST

def same_bsts(array_one, array_two)
  return are_same_bsts(array_one, array_two, 0, 0, -Float::INFINITY, Float::INFINITY)
end

def are_same_bsts(array_one, array_two, root_idx_one, root_idx_two, min_val, max_val)
  if root_idx_one == -1 || root_idx_two == -1
    return root_idx_one == root_idx_two
  end

  if array_one[root_idx_one] != array_two[root_idx_two]
    return false
  end

  left_root_idx_one = get_idx_of_first_smaller(array_one, root_idx_one, min_val)
  left_root_idx_two = get_idx_of_first_smaller(array_two, root_idx_two, min_val)
  right_root_idx_one = get_idx_of_first_bigger_or_equal(array_one, root_idx_one, max_val)
  right_root_idx_two = get_idx_of_first_bigger_or_equal(array_two, root_idx_two, max_val)

  current_value = array_one[root_idx_one]
  left_are_same = are_same_bsts(array_one, array_two, left_root_idx_one, left_root_idx_two, min_val, current_value)
  right_are_same = are_same_bsts(array_one, array_two, right_root_idx_one, right_root_idx_two, current_value, max_val)

  return left_are_same && right_are_same
end

def get_idx_of_first_smaller(arr, starting_idx, min_val)
  (starting_idx + 1..arr.length - 1).each do |i|
    if arr[i] < arr[starting_idx] && arr[i] >= min_val
      return i
    end
  end
  return -1
end

def get_idx_of_first_bigger_or_equal(arr, starting_idx, max_val)
  (starting_idx + 1..arr.length - 1).each do |i|
    if arr[i] >= arr[starting_idx] && arr[i] < max_val
      return i
    end
  end
  return -1
end

File: 073SameBSTs/sameBsts.js
// Solution 1
// O(n^2) time / O(n^2) space

const sameBsts = (arrayOne, arrayTwo) => {
  if (arrayOne.length !== arrayTwo.length) return false;

  if (arrayOne.length === 0 && arrayTwo.length === 0) return true;

  if (arrayOne[0] !== arrayTwo[0]) return false;

  const leftOne = getSmaller(arrayOne);
  const leftTwo = getSmaller(arrayTwo);
  const rightOne = getBiggerOrEqual(arrayOne);
  const rightTwo = getBiggerOrEqual(arrayTwo);

  return sameBsts(leftOne, leftTwo) && sameBsts(rightOne, rightTwo);
};

const getSmaller = (array) => {
  const smaller = [];
  for (let i = 1; i < array.length; i++) {
    if (array[i] < array[0]) smaller.push(array[i]);
  }
  return smaller;
};

const getBiggerOrEqual = (array) => {
  const biggerOrEqual = [];
  for (let i = 1; i < array.length; i++) {
    if (array[i] >= array[0]) biggerOrEqual.push(array[i]);
  }
  return biggerOrEqual;
};

// Solution 2
// O(n^2) time / O(d) space
// d - depth of the BST

const sameBsts = (arrayOne, arrayTwo) => {
  return areSameBsts(arrayOne, arrayTwo, 0, 0, -Infinity, Infinity);
};

const areSameBsts = (
  arrayOne,
  arrayTwo,
  rootIdxOne,
  rootIdxTwo,
  minVal,
  maxVal
) => {
  if (rootIdxOne === -1 || rootIdxTwo === -1) return rootIdxOne === rootIdxTwo;

  if (arrayOne[rootIdxOne] !== arrayTwo[rootIdxTwo]) return false;

  const leftRootIdxOne = getIdxOfFirstSmaller(arrayOne, rootIdxOne, minVal);
  const leftRootIdxTwo = getIdxOfFirstSmaller(arrayTwo, rootIdxTwo, minVal);
  const rightRootIdxOne = getIdxOfFirstBiggerOrEqual(
    arrayOne,
    rootIdxOne,
    maxVal
  );
  const rightRootIdxTwo = getIdxOfFirstBiggerOrEqual(
    arrayTwo,
    rootIdxTwo,
    maxVal
  );

  const currentValue = arrayOne[rootIdxOne];
  const leftAreSame = areSameBsts(
    arrayOne,
    arrayTwo,
    leftRootIdxOne,
    leftRootIdxTwo,
    minVal,
    currentValue
  );
  const rightAreSame = areSameBsts(
    arrayOne,
    arrayTwo,
    rightRootIdxOne,
    rightRootIdxTwo,
    currentValue,
    maxVal
  );

  return leftAreSame && rightAreSame;
};

const getIdxOfFirstSmaller = (array, startingIdx, minVal) => {
  for (let i = startingIdx + 1; i < array.length; i++) {
    if (array[i] < array[startingIdx] && array[i] >= minVal) return i;
  }
  return -1;
};

const getIdxOfFirstBiggerOrEqual = (array, startingIdx, maxValue) => {
  for (let i = startingIdx + 1; i < array.length; i++) {
    if (array[i] >= array[startingIdx] && array[i] < maxValue) return i;
  }
  return -1;
};
